go/vt/vtgate/planbuilder/operators/fk_verify.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2023 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package operators
+
+import (
+	"vitess.io/vitess/go/vt/sqlparser"
+	"vitess.io/vitess/go/vt/vtgate/planbuilder/operators/ops"
+	"vitess.io/vitess/go/vt/vtgate/planbuilder/plancontext"
+	"vitess.io/vitess/go/vt/vtgate/semantics"
+	"vitess.io/vitess/go/vt/vtgate/vindexes"
+)
+
+// FKVerify represents an insert operation on a table.
+type FKVerify struct {
+	// Sources are used to run the validation queries.
+	Sources      []ops.Operator
+	InputOffsets [][]int
+
+	// ParentFKs to verify during insert planning
+	ParentFKs []vindexes.ParentFKInfo
+
+	noColumns
+	noPredicates
+}
+
+func (fkVerify *FKVerify) Inputs() []ops.Operator {
+	return fkVerify.Sources
+}
+
+func (fkVerify *FKVerify) SetInputs(operators []ops.Operator) {
+	fkVerify.Sources = operators
+}
+
+func (fkVerify *FKVerify) ShortDescription() string {
+	return "FKVerify"
+}
+
+func (fkVerify *FKVerify) GetOrdering() ([]ops.OrderBy, error) {
+	panic("does not expect insert operator to receive get ordering call")
+}
+
+// Clone will return a copy of this operator, protected so changed to the original will not impact the clone
+func (fkVerify *FKVerify) Clone(inputs []ops.Operator) ops.Operator {
+	return &FKVerify{
+		Sources: inputs,
+	}
+}
+
+// NewFkVerify creates a new FkVerify operator given the parent foreign key validations we want it to run.
+func NewFkVerify(ctx *plancontext.PlanningContext, parentFKs []vindexes.ParentFKInfo, insertColumns sqlparser.Columns) (*FKVerify, error) {
+	var verifyQueryOps []ops.Operator
+	var inputOffsets [][]int
+	for _, fk := range parentFKs {
+		// Build the query for verification
```

Comment:
why not just parse here? this seems more fragile for no apparent benefit

Reply:
building a string query can sometimes miss quotes for the keywords. Building an AST is safer.